### PR TITLE
[NTP] 🐞 Fix config template to init default parameters

### LINKED ISSUE #17906: The NTP_SERVER configuration generated from the minigraph doesn't meet the new schema requirements.
#### Description

The `NTP_SERVER` configuration generated from the minigraph doesn't meet the new schema requirements.

The issue is caused by the changes introduced in [[NTP] Add NTP extended configuration](https://github.com/sonic-net/sonic-buildimage/pull/15058) PR. The PR updates the config DB schema and makes the NTP configuration generated from the minigraph incompatible.  

#### Steps to reproduce the issue:
1. Add NTP servers configuration to the minigraph file:
```XML
          <a:DeviceProperty>
            <a:Name>NtpResources</a:Name>
            <a:Reference i:nil="true"/>
            <a:Value>10.210.25.32;10.75.202.2</a:Value>
          </a:DeviceProperty>
```
2. Run sonic-cfggen utility generate config DB configuration from the minigraph:
```
sonic-cfggen -m /etc/sonic/minigraph.xml  --print-data
```

#### Describe the results you received:
The utility generates the NTP_SERVER configuration in the following format:
```
    "NTP_SERVER": {
        "10.210.25.32": {},
        "10.75.202.2": {}
    },
```

As a result, the NTP feature ignores this configuration, and the time synchronization doesn't work:
```
# show ntp
MGMT_VRF_CONFIG is not present.
unsynchronised
   polling server every 1 s
```

#### Describe the results you expected:
The NTP feature implementation expects to have an extended configuration in the format like this:
```
    "NTP_SERVER": {
        "internal_ntp_server1": {
            "association_type": "server",
            "iburst": "on",
            "admin_state": "enabled",
            "version": 3,
            "resolve_as": "10.210.25.32"
       },
        "internal_ntp_server2": {
            "association_type": "server",
            "iburst": "on",
            "admin_state": "enabled",
            "version": 3,
            "resolve_as": "10.75.202.2"
       },
    },
```
The YANG model for the configuration is defined here [sonic-ntp.yang](https://github.com/sonic-net/sonic-buildimage/blob/master/src/sonic-yang-models/yang-models/sonic-ntp.yang#L163) 

#### Output of `show version`:

```
# show version

SONiC Software Version: SONiC.master.560-c274be2e5_Internal
SONiC OS Version: 12
Distribution: Debian 12.4
Kernel: 6.1.0-11-2-amd64
Build commit: c274be2e5
Build date: Mon Jan 22 21:57:27 UTC 2024
Built by: sw-r2d2-bot@r-build-sonic-ci02-244

Platform: x86_64-mlnx_msn4700-r0
HwSKU: Mellanox-SN4700-O28
ASIC: mellanox
ASIC Count: 1
Serial Number: MT2247XZ0588
Model Number: MSN4700-WS2F
Hardware Revision: A1
Uptime: 02:17:59 up  1:25,  1 user,  load average: 0.41, 0.32, 0.35
Date: Fri 26 Jan 2024 02:17:59
```
#### Output of `show techsupport`:

```
(paste your output here or download and attach the file here )
```

### PR DESCRIPTION
**Pull request by fastiuk**
fixes #17906

#### Why I did it
To fix NTP config generation from the minigraph and save backward compatability

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it
Align `ntp.conf.j2` template to generate config out of empty `NTP_SERVER` DB configuration

#### How to verify it
Out of that NTP_SERVER configuration:
```json
{
  "10.210.25.32": {},
  "10.75.202.2": {}
}
```

The next config in `ntp.conf` file should be produced:
```
server 10.210.25.32
restrict 10.210.25.32 kod limited nomodify notrap noquery nopeer

server 10.75.202.2
restrict 10.75.202.2 kod limited nomodify notrap noquery nopeer
```

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305
- [x] 202311

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ] 

#### A picture of a cute animal (it is my cat Finn)
PXL_20230413_140941610 PORTRAIT

### CODE REVIEW

**Comment by qiluo-msft** [comment]
You mentioned "The NTP_SERVER configuration generated from the minigraph doesn't meet the new schema requirements" in the issue. To solve the problem, it is better to fix the schema (https://github.com/sonic-net/sonic-buildimage/blob/master/src/sonic-yang-models/yang-models/sonic-ntp.yang) instead of fixing the config generation behavior. The behavior is old from 2017. Any new schema design should be backward-compatible.

**Review by qiluo-msft — CHANGES_REQUESTED** [review]
Block as above comment.

**Comment by ganglyu** [comment]
Please check db_migrator, https://github.com/sonic-net/sonic-utilities/blob/master/scripts/db_migrator.py
Do we need to update db_migrator to migrate old ntp config to new schema?

**Comment by fastiuk** [comment]
> You mentioned "The NTP_SERVER configuration generated from the minigraph doesn't meet the new schema requirements" in the issue. To solve the problem, it is better to fix the schema (https://github.com/sonic-net/sonic-buildimage/blob/master/src/sonic-yang-models/yang-models/sonic-ntp.yang) instead of fixing the config generation behavior. The behavior is old from 2017. Any new schema design should be backward-compatible.

Minigraph doesn't use YANG to generate config, because there are default values in YANG. Anyway I already changed ntp.conf.j2 template according to your request, so it will preserve backward compatibility

**Comment by fastiuk** [comment]
> Please check db_migrator, https://github.com/sonic-net/sonic-utilities/blob/master/scripts/db_migrator.py Do we need to update db_migrator to migrate old ntp config to new schema?

No need according to changes @qiluo-msft requested

**Comment by fastiuk** [comment]
The PR checker was failed because ntpsec does not support `nopeer` flag anymore: https://docs.ntpsec.org/latest/accopt.html
Fixed it by removing nopeer configuration.

**Comment by anamehra** [comment]
How do we configure some of the config params like iburst for NTP server used in this j2 file via sonic-mgmt? Is there any document to refer to?

**Comment by liat-grozovik** [comment]
@oleksandrivantsiv @dgsudharsan could you please help to review this fix?

File: files/image_config/ntp/ntp.conf.j2 [hunk 1]
```diff
@@ -47,17 +47,13 @@ config -#}
...
-    {% else -%}
-        {% set aoptions = aoptions ~ ' nopeer' -%}
+    {% if association_type == 'pool' -%}
+        {% set resolve_as = server -%}
     {% endif -%}
 
-{{ config.association_type }} {{ config_as }}{{ soptions }}
+{{ association_type }} {{ resolve_as }}{{ soptions }}
 {% if global.server_role == 'disabled' %}
-restrict {{ config_as }} kod limited nomodify notrap noquery{{ aoptions }}
```
Comment (dgsudharsan) [inline files/image_config/ntp/ntp.conf.j2:60]: Can you please clarify why notrap and aoptions are removed?

Comment (fastiuk) [reply files/image_config/ntp/ntp.conf.j2:60]: `nopeer` and `notrap` is not supported by ntp-sec anymore. You can check its reference here: https://docs.ntpsec.org/latest/accopt.html
`aoptions` was used to add `nopeer` based on condition (if server is not a pool), but now it is not needed anymore.

**Comment by fastiuk** [comment]
> How do we configure some of the config params like config.iburst, config.resolve_as, etc. for NTP server used in this j2 file via sonic-mgmt? Is there any document to refer to?

`config.iburst` only trough the DB. So far only server can be added via SONiC CLI. `config.resolve_as` is internal option for a NTP server and shouldn't be configured directly. It used to save resolved address, because when you configure `time.google.com` for example it changes address every time. So now `config.resolve_as` is configured via SONiC CLI when new NTP server is added.

**Comment by fastiuk** [comment]
> @oleksandrivantsiv @dgsudharsan could you please help to review this fix?

@oleksandrivantsiv @dgsudharsan : checkers are passed. All comments handled, please approve if looks good

**Review by oleksandrivantsiv — APPROVED** [review]

File: files/image_config/ntp/ntp.conf.j2 (hunk 1 above)
Comment (dgsudharsan) [reply files/image_config/ntp/ntp.conf.j2:60]: @fastiuk Then we may need to raise a separate PR for 202311 as 202311 doesn't use bookworm. @saiarcot895 FYI

**Review by dgsudharsan — APPROVED** [review]

**Review by anamehra — APPROVED** [review]

**Review by saiarcot895 — APPROVED** [review]

**Comment by fastiuk** [comment]
@qiluo-msft the PR was approved. Could you please merge it?

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**